pkg/worker/worker.go

```diff
@@ -0,0 +1,300 @@
+/*
+Package worker allows to start, stop and query Linux processes as concurrent jobs.
+
+Each job is saved to a provided storage via DataStorage interface. Its progress
+is updated each time a new line in StdIn or StdErr appears.
+
+To make use of the package create a new Worker using its constructor - NewWorker
+and provide a correct DataStorage. Control the processes using Start() and Stop()
+methods. Query the output anytime, even when the process is running using Query()
+method.
+
+DataStorage interface must include working data synchronization methods: Lock()
+and Unlock().
+*/
+package worker
+
+import (
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/google/uuid"
+)
+
+var (
+	// ErrJobNotFound indicates that the requested job
+	// has not been found in the DataStorage.
+	ErrJobNotFound = errors.New("job has not been found")
+
+	// ErrJobAlreadyStopped indicates that the requested job
+	// has already been stopped.
+	ErrJobAlreadyStopped = errors.New("job has already been stopped")
+
+	// ErrJobAlreadyFinished indicates that the requested job
+	// has already been finished.
+	ErrJobAlreadyFinished = errors.New("job has already been finished")
+)
+
+// DataStorage allows to store and retrieve jobs from any storage that implements the interface.
+// The storage needs to implement sync.Mutex or other form of access synchronization based on
+// Lock() and Unlock() methods.
+type DataStorage interface {
+	Store(job *Job) error
+	Retrieve(uuid uuid.UUID) (*Job, error)
+}
+
+// MemoryStorage is a simple, in-memory implementation of the DataStorage interface.
+// Avoid using in production. Implement DataStorage with database of your choice.
+type MemoryStorage struct {
+	Data  map[uuid.UUID]*Job
+	mutex sync.Mutex
+}
+
+// Store saves the job in memory.
+func (m *MemoryStorage) Store(job *Job) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	logs := job.Logs()
+	job.SetLogs(logs)
+	m.Data[job.UUID] = job
+	return nil
+}
+
+// Retrieve loads the job from the memory storage.
+func (m *MemoryStorage) Retrieve(uuid uuid.UUID) (*Job, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	job, ok := m.Data[uuid]
+	if !ok {
+		return nil, ErrJobNotFound
+	}
+	logs := job.Logs()
+	job.SetLogs(logs)
```

Comment:
Sorry I'm not sure what is the point of retrieving the job's logs and then setting them again? Same above.

Reply:
This was my poor attempt to slice deep copy. The slice was copied inside the helper functions, but it was a bad solution. I changed the logs type to Buffer and removed this logic.